src/include/sof/audio/format.h

```diff
@@ -79,6 +79,13 @@
 #define SATP_INT32(x) (((x) > INT32_MAX) ? INT32_MAX : (x))
 #define SATM_INT32(x) (((x) < INT32_MIN) ? INT32_MIN : (x))
 
+/* Arithmetic shift right */
+#define ARITHMETICSHIFTRIGHT(u, n) ({	\
+	(u > 0) ?	\
+		(int32_t)((uint32_t)(((uint32_t)u) >> ((uint64_t)n))) :	\
```

Comment:
Why is the shift right amount cast to 64 bit type?

Reply:
I guess the shifted value itself shall be cast to 64-bit? One of them needs to be 64-bit for the result to work in 64-bit domain.

Reply:
This looks wrong and may generate inefficient code on 32bit devices. I would also change this to be a static inline function since the types are important here. e.g. (and if I understand the usage correctly)
```c
static inline uint32_t arith_shift_right(uint32_t u, uint32_t n)
{
    if (u > 0)
        return u >> n;
    else
        return abs(u) >> n;
}
```
We will also need an abs().

Reply:
@lgirdwood , Thanks for  your proposal. I think with your proposal- msb bit will be lost i.e. the  sign bit
For Example : 

```
 +---+----------+--------+---------+
 |n  |  u       |   y    |   y_liam|
 +---+----------+--------+---------+
 |4  | -50940067|-3183755|  3183754|
 +---+----------+--------+---------+

```

**In matlab -** 
bitsra(-50940067,4)
ans =
-3183754.1875

Test result with few sample data on signed bit

                                            =========================================

                                            =========================================
                                            ---Aithmetic shift right on signed bit---
                                            =========================================

                                            =========================================
           iter |   u-operand(i) |       nbits(i) |     y-outpt(i) |    Matlab-bsra |    c-out-fixed |      error-ppm |
           -----+----------------+----------------+----------------+----------------+----------------+----------------+
              1 |     -652032874 |              1 |  -3.260164e+08 |     -326016437 |     -326016437 |       0.000000 |
              2 |     -326016437 |              2 |      -81504110 |  -8.150411e+07 |      -81504110 |       0.000000 |
              3 |      -81504110 |              3 |  -1.018801e+07 |  -1.018801e+07 |      -10188014 |       0.000000 |
              4 |      -15918772 |              5 |  -4.974626e+05 |  -4.974616e+05 |        -497462 |       0.000000 |
              5 |       -7550352 |              8 |  -2.949456e+04 |  -2.949356e+04 |         -29494 |       0.000000 |
              6 |      -11744509 |              9 |  -2.293949e+04 |  -2.293849e+04 |         -22939 |       0.000000 |
              7 |      -13841530 |             10 |  -1.351812e+04 |  -1.351712e+04 |         -13518 |       0.000000 |
              8 |      -12793042 |             11 |  -6.247602e+03 |  -6.246603e+03 |          -6247 |       0.000000 |
              9 |      -13317292 |             12 |  -3.252292e+03 |  -3.251292e+03 |          -3252 |       0.000000 |
             10 |      -13055169 |             13 |  -1.594648e+03 |  -1.593649e+03 |          -1594 |       0.000000 |
             11 |      -13186231 |             14 |  -8.058236e+02 |  -8.048237e+02 |           -805 |       0.000000 |
             12 |      -13120700 |             15 |  -4.014120e+02 |  -4.004120e+02 |           -401 |       0.000000 |
             13 |      -13153465 |             16 |  -2.017059e+02 |  -2.007059e+02 |           -201 |       0.000000 |
             14 |      -13169847 |             17 |  -1.014780e+02 |  -1.004780e+02 |           -101 |       0.000000 |
             15 |      -13178038 |             18 |  -5.127022e+01 |  -5.027023e+01 |            -51 |       0.000000 |
             16 |      -13173943 |             19 |  -2.612730e+01 |  -2.512730e+01 |            -26 |       0.000000 |
             17 |      -13175990 |             20 |  -1.356560e+01 |  -1.256560e+01 |            -13 |       0.000000 |
             18 |      -13177013 |             21 |  -7.283289e+00 |  -6.283289e+00 |             -7 |       0.000000 |
             19 |      -13176502 |             22 |  -4.141523e+00 |  -3.141523e+00 |             -4 |       0.000000 |
             20 |      -13176757 |             23 |  -2.570792e+00 |  -1.570792e+00 |             -2 |       0.000000 |
             21 |      -13176884 |             24 |  -1.785403e+00 |  -7.854035e-01 |             -1 |       0.000001 |